packages/core/src/indexing/service.ts

```diff
@@ -1,103 +1,134 @@
+  buildDB,
+  buildNetwork,
+} from "./context.js";
 import { addUserStackTrace } from "./trace.js";
-import { ponderTransport } from "./transport.js";
 
 type IndexingEvents = {
   eventsProcessed: { toCheckpoint: Checkpoint };
   error: { error: Error };
 };
 
-type LogEvent = {
-  networkName: string;
-  contractName: string;
-  eventName: string;
-  chainId: number;
-  args: any;
-  log: Log;
-  block: Block;
-  transaction: Transaction;
-};
 type SetupTask = {
   kind: "SETUP";
   event: {
     networkName: string;
     contractName: string;
-    chainId: number;
-    blockNumber: number;
+    checkpoint: Checkpoint;
+  };
+};
+type LogEventTask = {
+  kind: "LOG";
+  event: {
+    networkName: string;
+    contractName: string;
+    eventName: string;
+    event: {
+      args: any;
+      log: Log;
+      block: Block;
+      transaction: Transaction;
+    };
+    checkpoint: Checkpoint;
+    endCheckpoint?: Checkpoint;
+    eventsProcessed?: number;
   };
 };
-type LogEventTask = { kind: "LOG"; event: LogEvent };
+
 type IndexingFunctionTask = SetupTask | LogEventTask;
 type IndexingFunctionQueue = Queue<IndexingFunctionTask>;
 
+// Note: this should move to a dynamic value, based on how many indexing function keys there are.
+const TASK_BATCH_SIZE = 1_000;
+
 export class IndexingService extends Emittery<IndexingEvents> {
   private common: Common;
   private indexingStore: IndexingStore;
   private syncGatewayService: SyncGateway;
   private sources: Source[];
-  private contexts: Record<
-    number,
-    {
-      client: Client;
-      network: { chainId: number; name: string };
-      contracts: Record<
-        string,
-        {
-          abi: Abi;
-          address?: Address | readonly Address[];
-          startBlock: number;
-          endBlock?: number;
-          maxBlockRange?: number;
-        }
-      >;
-    }
-  > = {};
+  private networks: Network[];
 
-  private schema?: Schema;
-  private db: Record<string, DatabaseModel<any>> = {};
+  private isPaused = false;
 
   private indexingFunctions?: IndexingFunctions;
+  private schema?: Schema;
+  private tableAccess?: TableAccess;
 
-  private eventProcessingMutex: Mutex;
-  private queue?: IndexingFunctionQueue;
+  queue?: IndexingFunctionQueue;
 
-  private eventsProcessedToCheckpoint: Checkpoint = zeroCheckpoint;
+  private getNetwork: (checkpoint: Checkpoint) => Context["network"] =
+    undefined!;
+  private getClient: (checkpoint: Checkpoint) => Context["client"] = undefined!;
+  private getDB: (checkpoint: Checkpoint) => Context["db"] = undefined!;
+  private getContracts: (checkpoint: Checkpoint) => Context["contracts"] =
+    undefined!;
 
-  private currentIndexingCheckpoint: Checkpoint = zeroCheckpoint;
-  private isPaused = false;
-  /** If true, failed tasks should not log errors or be retried. */
-  private isShuttingDown = false;
+  private indexingFunctionMap: Record<
+    /* Indexing function key: "{ContractName}:{EventName}" */
+    string,
```

Comment:
Wonder if it would be worth using branded types for some of these keys (rather than the comment). Would also be useful for self-documenting the `parents` field below, for example.